File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int i = 0, n = s.size(), res = 0, maxCount = 0;
        vector<int> charCount(26, 0);
        for(int j = 0; j < n; j++) {
            charCount[s[j] - 'A']++;
            maxCount = max(charCount[s[j] - 'A'], maxCount);
            // window size - maxCount yields minCount which should be greater than k to shift the window
            while(j - i - maxCount + 1 > k) { 
                charCount[s[i] - 'A']--;
                i++;
            }
            res = max(j - i + 1, res);
        }

        return res;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int n = s.length(), i = 0, maxFreq = 0, res = 0;
        int[] charFreq = new int[26];

        for (int j = 0; j < n; j++) {
            charFreq[s.charAt(j) - 'A']++;
            maxFreq = Math.max(maxFreq, charFreq[s.charAt(j) - 'A']);

            if (j - i + 1 - maxFreq > k) {
                charFreq[s.charAt(i) - 'A']--;
                i++;
            }

            res = Math.max(res, j - i + 1);
        }

        return res;
    }
}